drivers/serial/uart_stm32.c

```diff
@@ -39,6 +39,15 @@
 #include <stm32_ll_exti.h>
 #endif /* CONFIG_PM */
 
+#define _UART_STM32_EACH_NODE_HAS_PROP_OR(id, prop) DT_INST_NODE_HAS_PROP(id, prop) &&
+#define UART_STM32_EACH_NODE_HAS_PROP(prop) \
+	(DT_INST_FOREACH_STATUS_OKAY_VARGS(_UART_STM32_EACH_NODE_HAS_PROP_OR, prop) 1)
+
+#if !(defined(CONFIG_UART_USE_RUNTIME_CONFIGURE) || UART_STM32_EACH_NODE_HAS_PROP(current_speed))
+#	error "You need either enable CONFIG_UART_USE_RUNTIME_CONFIGURE" \
+	      "or define `current-speed` in each node"
```

Comment:
Note: that this whole point is valid only if USART Auto baud rate detection is not enabled (not implemented so far).

Then, I'm not quite convinced about this, since:
- Not setting baudrate in device tree is rather a trick for experienced users
- CONFIG_UART_USE_RUNTIME_CONFIGURE is enabled by default anyway

If someone ends up in this error, this is likely a rookie who doesn't know what he's doing.
Enabling CONFIG_UART_USE_RUNTIME_CONFIGURE will avoid the error but won't guarantee a safe behavior. If `uart_configure()` is not called to set baudrate on the uart instace on which baudrate is not set in dt, an issue will pops up anyway.

The only real way to catch a real problem would be to check if baudrate was once configured before starting a transaction (`data->baud_rate` not null). This would have to be implemented in each transaction function.
In the end that's some footprint overhead for a quite marginal clarification of the way to use the driver API.

If some change has to be made, I prefer to make baudrate required in the binding and provide some details in the description stating that `baudrate` property only describes the boot configuration of the device and for some cases (such as modbus), it will be updated by the application in runtime.